File: leetcode/1399-CountLargestGroup/solution.ts
function countLargestGroup(n: number): number {
    let hashmap = new Map();
    for (let i = 1; i<=n; i++){
        let sumOfDigits = 0;
        let j = i;
        while (j!=0){
            sumOfDigits += j%10;
            j = Math.floor(j/10);
        }
        if (hashmap.has(sumOfDigits))
            hashmap.set(sumOfDigits, hashmap.get(sumOfDigits)+1);
        else
            hashmap.set(sumOfDigits, 1);
    }
    // console.log(hashmap)
    let largestSize = Math.max(...hashmap.values());
    let groupsWithLargestSize = 0;
    
    for (const [sumDigits, size] of hashmap){
        if (size == largestSize)
            groupsWithLargestSize += 1
    }


    return groupsWithLargestSize;
};

File: leetcode/1399-CountLargestGroup/solution.py
from collections import defaultdict

class Solution:
    def countLargestGroup(self, n: int) -> int:
        # use dictionary with digit sum as key and occurence of the sum as value
        # iterate from 1 to n and fill dictionary, get max size, then get how many groups match the largest size
        # time: O(n), space: O(n)

        groups = defaultdict(int) # key = sum, value = group size
        # go through each number from 1 to n
        for i in range(1, n+1):
            sum_digits = 0
            # break down integer to digits, sum them
            while i:
                sum_digits += i % 10
                i //= 10
            # add to dictionary
            groups[sum_digits] += 1
        
        # get largest size
        largest_size = max(groups.values())
        num_largest_size = 0
        # count how many groups with largest size
        for group, size in groups.items():
            if size == largest_size:
                num_largest_size += 1
        return num_largest_size